arbor/label_resolution.cpp

```diff
@@ -0,0 +1,154 @@
+    if (size() < 1) return std::nullopt;
+    auto part = util::partition_view(ranges_partition);
+    // Index of the range containing idx.
+    auto ridx = part.index(idx);
+
+    // First element of the range containing idx.
+    const auto& start = ranges.at(ridx).begin;
+
+    // Offset into the range containing idx.
+    const auto& range_part = part.at(ridx);
+    auto offset = idx - range_part.first;
+
+    return start + offset;
+}
+
+const label_resolution_map::range_set& label_resolution_map::at(const cell_gid_type& gid, const cell_tag_type& tag) const {
+    return map.at(gid).at(tag);
+}
+
+std::size_t label_resolution_map::count(const cell_gid_type& gid, const cell_tag_type& tag) const {
+    if (!map.count(gid)) return 0u;
+    return map.at(gid).count(tag);
+}
+
+label_resolution_map::label_resolution_map(const cell_labels_and_gids& clg) {
+    arb_assert(clg.label_range.check_invariant());
+    const auto& gids = clg.gids;
+    const auto& labels = clg.label_range.labels();
+    const auto& ranges = clg.label_range.ranges();
+    const auto& sizes = clg.label_range.sizes();
+
+    std::vector<cell_size_type> label_divs;
+    auto partn = util::make_partition(label_divs, sizes);
+    for (auto i: util::count_along(partn)) {
+        auto gid = gids[i];
+
+        std::unordered_map<cell_tag_type, range_set> m;
+        for (auto label_idx: util::make_span(partn[i])) {
+            const auto range = ranges[label_idx];
+            auto size = int(range.end - range.begin);
+            if (size < 0) {
+                throw arb::arbor_internal_error("label_resolution_map: invalid lid_range");
+            }
+            auto& range_set = m[labels[label_idx]];
+            range_set.ranges.push_back(range);
+            range_set.ranges_partition.push_back(range_set.ranges_partition.back() + size);
+        }
+        if (!map.insert({gid, std::move(m)}).second) {
+            throw arb::arbor_internal_error("label_resolution_map: duplicate gid");
+        }
+    }
+}
+
+// resolver methods
+std::optional<cell_lid_type> resolver::resolve(const cell_global_label_type& iden, const label_resolution_map& label_map) {
+    if (!label_map.count(iden.gid, iden.label.tag)) {
+        throw arb::bad_connection_label(iden.gid, iden.label.tag);
+    }
+    const auto& range_set = label_map.at(iden.gid, iden.label.tag);
+    if (!range_set.size()) return std::nullopt;
+
+    switch (iden.label.policy) {
+    case lid_selection_policy::round_robin: {
+        // Get the state of the round_robin iterator.
+        auto& rr_state = state_map[iden.gid][iden.label.tag][iden.label.policy];
+        auto idx = std::get<round_robin_state>(rr_state).state;
+
+        // Update the state of the round_robin iterator.
+        rr_state = round_robin_state((idx+1) % range_set.size());
+
+        // Get the lid at the current index.
+        return range_set.at(idx);
+    }
```

Comment:
As an indentation thing, because these inner blocks demarcated by braces are one scope deeper than that of the switch statement, it'd be clearer if they were indented (for the same reason essentially that we _don't_ indent case labels in switches).

e.g.
```
switch (...) {
case foo:
    // in scope of 'switch' here.
    {
        // in scope of inner block here.
        ...
    }
case bar: // etc,
}
```